app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.test.js

```diff
@@ -0,0 +1,521 @@
+import { ethErrors } from 'eth-rpc-errors';
+import {
+  CHAIN_IDS,
+  NETWORK_TYPES,
+} from '../../../../../shared/constants/network';
+import addEthereumChain from './add-ethereum-chain';
+
+const NON_INFURA_CHAIN_ID = '0x123456789';
+
+const MOCK_MAINNET_CONFIGURATION = {
+  chainId: CHAIN_IDS.MAINNET,
+  nickname: 'Ethereum Mainnet',
+  rpcUrl: 'https://mainnet.infura.io/v3/',
+  type: NETWORK_TYPES.MAINNET,
+  ticker: 'ETH',
+  rpcPrefs: {
+    blockExplorerUrl: 'https://etherscan.io',
+  },
+};
+
+const MOCK_OPTIMISM_CONFIGURATION = {
+  chainId: CHAIN_IDS.OPTIMISM,
+  nickname: 'Optimism',
+  rpcUrl: 'https://optimism.llamarpc.com',
+  rpcPrefs: {
+    blockExplorerUrl: 'https://optimistic.etherscan.io',
+  },
+  ticker: 'ETH',
+};
+
+const MOCK_NON_INFURA_CONFIGURATION = {
+  chainId: NON_INFURA_CHAIN_ID,
+  rpcUrl: 'https://custom.network',
+  ticker: 'CUST',
+  nickname: 'Custom Network',
+  rpcPrefs: {
+    blockExplorerUrl: 'https://custom.blockexplorer',
+  },
+};
```

Comment:
Consider replacing these with factory functions so that we can preclude the possibility of breaking test isolation.

Reply:
We could also just `Object.freeze()` them instead?

Reply:
We'd want to deep freeze them if so, since they contain nested objects. Easier to just wrap them like `const getThing = () => ({ ... });` and then just spam `getThing()` ad nauseam.

Reply:
Done here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/85dbab821eb7f179d253b7926d4c858be3388837

I deduped the calls so we just call the factory function once per test.